hugr-core/src/types/type_param.rs

```diff
@@ -383,29 +388,33 @@ pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgErr
             _,
         ) if param.contains(cached_decl) => Ok(()),
         (TypeArg::Type { ty }, TypeParam::Type { b: bound })
-            if bound.contains(ty.least_upper_bound()) && !ty.is_row_var() =>
+            if bound.contains(ty.least_upper_bound()) =>
         {
             Ok(())
         }
         (TypeArg::Sequence { elems }, TypeParam::List { param }) => {
             elems.iter().try_for_each(|arg| {
                 // Also allow elements that are RowVars if fitting into a List of Types
-                if let (TypeArg::Type { ty }, TypeParam::Type { b }) = (arg, &**param) {
-                    if ty.is_row_var() && b.contains(ty.least_upper_bound()) {
-                        return Ok(());
+                if let (
+                    TypeArg::Variable {
+                        v:
+                            TypeArgVariable {
+                                idx: _,
+                                cached_decl: TypeParam::List { param: arg_var },
+                            },
+                    },
+                    TypeParam::Type { b: param_bound },
+                ) = (arg, &**param)
+                {
+                    if let TypeParam::Type { b: var_bound } = **arg_var {
+                        if param_bound.contains(var_bound) {
+                            return Ok(());
+                        }
                     }
                 }
                 check_type_arg(arg, param)
             })
         }
-        // Also allow a single "Type" to be used for a List *only* if the Type is a row variable
```

Comment:
This case is now handled by the standard case for the TypeArg being a Variable :)